Description: Enable GPS by enabling flags in Makefile
Author: Iain R. Learmonth <irl@fsfe.org>                                         
Last-Update: 2015-02-05

--- direwolf.orig/Makefile.linux
+++ direwolf/Makefile.linux
@@ -100,8 +100,8 @@
 
 # Uncomment following lines to enable GPS interface & tracker function.
 
-#CFLAGS += -DENABLE_GPS
-#LDLIBS += -lgps
+CFLAGS += -DENABLE_GPS
+LDLIBS += -lgps
 
 
 
