--- a/Makefile.linux
+++ b/Makefile.linux
@@ -447,7 +447,7 @@
 # Multiple AGWPE network or serial port clients to test TNCs side by side.
 
 aclients : aclients.c ax25_pad.c fcs_calc.c textcolor.o misc.a
-	$(CC) $(CFLAGS) -g -o $@ $^ 
+	$(CC) $(CFLAGS) -g -o $@ $^ $(LDFLAGS)
 
 
 # Talk to a KISS TNC.
@@ -466,7 +466,7 @@
 # Touch Tone to Speech sample application.
 
 ttcalc : ttcalc.o ax25_pad.o fcs_calc.o textcolor.o misc.a
-	$(CC) $(CFLAGS) -g -o $@ $^ 
+	$(CC) $(CFLAGS) -g -o $@ $^ $(LDFLAGS)
 
 
 # -----------------------------------------  Libraries  --------------------------------------------
